### PR TITLE
Free allocated memory in unpack_action_data

### PR DESCRIPTION
**Pull request by andriantolie**
Looking at issue https://github.com/EOSIO/eos/issues/4930 in these past days, I notice some people are still using `unpack_action_data` instead of `EOSIO_ABI.`

When using `unpack_action_data` and their action structure contains string, it will throw "env.free unresolveable" when uploading the contract. I guess it's not linked with the `free` defined in eosiolib.cpp?

### CODE REVIEW

File: contracts/eosiolib/action.hpp [hunk 1]
```diff
@@ -50,7 +50,11 @@ namespace eosio {
       size_t size = action_data_size();
       char* buffer = (char*)( max_stack_buffer_size < size ? malloc(size) : alloca(size) );
```
Comment (brianjohnson5972) [inline contracts/eosiolib/action.hpp]: Add:
const bool heap_allocation = max_stack_buffer_size < size;
and use heap_allocation in both places.

Comment (andriantolie) [reply contracts/eosiolib/action.hpp]: Thanks for the suggestion. It makes it clearer to read :)
I've updated the PR

**Comment by heifner** [comment]
@andriantolie I fail to see how this would have anything to do with "env.free unresolveable". Where you able to reproduce the "env.free unresolveable" error? And if so, did this fix it?

**Comment by andriantolie** [comment]
Yes, this actually doesn't really solve "env.free unresolvable" entirely. But this just solves the "env.free unresolvable" problem when people creates a simple contract with just unpack_action_data.

I couldn't figure out the reasoning behind this. But seems people get "env.free unresolvable" error when they include string object inside his action struct, but never called "free" function anywhere in the contract. That's why when people use EOSIO_ABI, or when they call "free" function at least once in the contract, they don't get the error. In the .wast file, the "free" function turns from "import" to "export" when it is being used in the contract. I don't have sufficient knowledge to understand the reasoning behind this though, so if somebody can explain I'm really happy to learn....

**Comment by heifner** [comment]
@larryk85 @spoonincode Is there someway we can tell wast to export instead of import to solve this unresolvable error without having to actually call free? The memory will all be free'd at the end of the action, so seems like a waste to free when it is not necessary.

**Pull request merged**